src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala

```diff
@@ -98,10 +122,60 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
   private val declined = mutable.TreeMap[ModifierId, Long]()
 
   /**
-    * The node stops to accept transactions if declined table reaches this max size. It prevents spam attacks trying
-    * to bloat the table (or exhaust node's CPU)
+    * Counter which contains total cost of transactions entered mempool or rejected by it since last block processed.
+    * Used to avoid sudden spikes in load, limiting transactions processing time and make it comparable to block's
+    * processing time
     */
-  private val MaxDeclined = 400
+  private var globalInfo = IncomingTxInfo.empty()
+
+  /**
+    * Cache which contains bytes of transactions we received but not parsed and processed yet
+    */
+  private val txProcessingCache = mutable.Map[ModifierId, TransactionProcessingCacheRecord]()
+
+  /**
+    * To be called when the node is synced and new block arrives, to reset transactions cost counter
+    */
+  private def clearTransactionsInfo(): Unit = {
+    globalInfo = IncomingTxInfo.empty()
+  }
+
+  /**
+    * To be called when the node is synced and new block arrives, to resume transaction bytes cache processing
+    */
+  private def processFirstTxProcessingCacheRecord(): Unit = {
+    txProcessingCache.headOption.foreach { case (txId, processingCacheRecord) =>
+      parseAndProcessTransaction(txId, processingCacheRecord.txBytes, processingCacheRecord.source)
+      txProcessingCache -= txId
+    }
+  }
+
+  /**
+    * To be called when mempool reporting on finished transaction validation.
+    * This method adds validation cost to counter and send another
+    */
+  private def processMempoolResult(processingResult: InitialTransactionCheckOutcome): Unit = {
+    val ReserveCostValue = 5000
+
+    val costOpt = processingResult.transaction.lastCost
+    if (costOpt.isEmpty) {
+      // should not be here, and so ReserveCostValue should not be used
+      log.warn("Cost is empty in processMempoolResult")
+    }
+    val cost = costOpt.getOrElse(ReserveCostValue)
+    val ng = processingResult match {
+      case _: FailedTransaction => globalInfo.copy(invalidatedCost = globalInfo.invalidatedCost + cost)
+      case _: SuccessfulTransaction => globalInfo.copy(acceptedCost = globalInfo.acceptedCost + cost)
+      case _: DeclinedTransaction => globalInfo.copy(declinedCost = globalInfo.declinedCost + cost)
+    }
+
+    if (globalInfo.totalCost < MempoolCyclesPerBlock) {
```

Comment:
Maybe `ng` should be used here instead of globalInfo?

Reply:
yeah, fixed